### PR TITLE
Make `find` faster

### PR DESCRIPTION
**Pull request by glacambre**
I'm opening this PR to talk about making `find` faster without spamming users subscribed to https://github.com/cmcaine/tridactyl/issues/64 .

A good website to test `find`'s speed is [this one](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html).

This [first commit](https://github.com/glacambre/tridactyl/commit/efee378d7bcab041aa50cff73b6fd8ccad03a5c6) removes an unnecessary filter operation. I believe it is unnecessary because `DOM.getElemsBySelector("*", ...)` will return the `<html></html>` element, which contains all other elements, including the filtered ones. On the cuda website, I couldn't measure any difference in speed between using `DOM.getElemesBySelector("*", ...)` and `[document.body]`.

I think there are two ways to improve `find`'s performance. The first one is to make Mark.js asynchronous. This might make things a bit complicated when dealing with updates (e.g. user types first letter -> we start finding matches -> user types a second letter but we haven't finished finding all of the first matches). Also, I took a quick glance at Mark.js's source code and found it a bit hard to understand so making it asynchronous without introducing new bugs might be hard.

The second option is to write our own find method in something that compiles to WASM (Rust ?). I already have an idea of how it could be implemented, but it would make the build process much slower and include a dependency on a whole new toolchain. Also, there's no guarantee that this new implementation would be fast enough to avoid needing to be made asynchronous.

What are your thoughts about this? Are there other options I haven't thought about?

### CODE REVIEW

**Comment by bovine3dom** [comment]
I haven't looked at the code in ages, but I think we were searching through more than just the text on the page. I remember on GitHub it was finding things that weren't there.

I haven't profiled the code at all (is that even possible?) so I don't really know which bits are slow.

**Comment by glacambre** [comment]
>I haven't profiled the code at all (is that even possible?)

Sort of. There's a profiler in the console (Shift-F5 by default) but it won't tell you much about the content script aside from when the GC is being triggered. The profiler for the background script is way more detailed but doesn't have any data about what's happening in the content script.

**Comment by bovine3dom** [comment]
https://github.com/cmcaine/tridactyl/commit/1441682eff9a150e2a703fad318a775e0bc64dac was my last attempt at fixing this - maybe you can decipher my ramblings?

**Comment by glacambre** [comment]
I already looked at your code, the reason filtering doesn't work [here](https://github.com/cmcaine/tridactyl/commit/1441682eff9a150e2a703fad318a775e0bc64dac#diff-65e210f052ab6c715a4344bbfff3c216R37) is that elemswithtext() returns the a list of elements that contains the body of the page and Mark.js will look at all its child nodes.
I believe if we try to filter elements before giving them to Mark.js we might as well re-implement Mark.js ourselves because a lot of the processing we'll be doing will be done by Mark.js too.

**Comment by bovine3dom** [comment]
https://en.wikipedia.org/wiki/Battle_of_Mosul_(2016%E2%80%932017) is a nice example of a page on which `find` is unusable.

I really doubt that JS is what's making it slow - JavaScript is actually really quick these days as companies have poured a fortune into optimising it, so I think switching to WASM is a bad idea.

I think the main problem is just that our algorithm is bad. Switching to async and stopping the current search if another character is typed is certainly sensible, but I think that should come later.

cVim, for example, still struggles on the Mosul page but is many orders of magnitude better. Its searching code starts here: https://github.com/1995eaton/chromium-vim/blob/2ed2c90944ebf8238d3e30ce282b2a9be3bdcdd1/content_scripts/mappings.js#L655

**Comment by glacambre** [comment]
I believe we can just throw everything away and use firefox's find API: https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/find/find

**Comment by bovine3dom** [comment]
We wanted regex search. I guess we could use find and then fall back to our slow implementation?

**Comment by bovine3dom** [comment]
Huh. Apparently not. cVim definitely has it, though...

**Comment by bovine3dom** [comment]
I'm getting more and more fed up by GitHub. If you make it work with Firefox's built in find, we can worry about regex later.

I'm still reluctant to merge anything that doesn't let people know what they're searching for. We could potentially just hack something in like the mode indicator that displays what is being searched for and is not a real text field.

**Comment by glacambre** [comment]
Or we could even hack it into the completion mechanism. / would just be `fillcmdline find ` and when you start typing letters the completion buffer is filled with candidates from the page. That's one less mode to worry about and I believe seeing all matches in the completion buffer could be pretty cool.

**Comment by bovine3dom** [comment]
Sure, that sounds neat, especially if we can display some context around the matches.

**Comment by glacambre** [comment]
Here's a beginning of proof of concept. You can't actually jump to words yet but you can have completions. Finding is extremely fast compared to what's in master right now. I'll continue working on this tomorrow.

**Comment by glacambre** [comment]
b8f7d5f Makes `:find` usable if you disable the completions (`set findresults 0`). If you don't do that, you'll hint a bug in the completion mechanism where pressing "enter" before the completions are initialized.
This might bug as https://github.com/cmcaine/tridactyl/issues/580 since the error message is also that `exstr` is undefined.

Also, I need to remember to document what I wrote.

**Comment by glacambre** [comment]
I've disabled completions. Find is now in a somewhat usable state, it's very fast even on large websites. There currently are two bugs, the first one is that elements that are not visible to the user will still be selected (this happens with the changelog on the newtab page) and the second one is that jumping from an element to another does not change the color of the selected element.

I started fixing the completions but this required to move some of the things sitting in commandline_background to the content script which in turn required a lot of other modifications that would have made other PRs very hard to rebase (or mine very hard to rebase once other PRs are merged) so this will probably have to wait until v2.

**Comment by glacambre** [comment]
I think I'm done.

Find works well on simple websites (e.g wikipedia) but not much on the ones that do complicated things (e.g. on bbc.com the current element is not always correctly highlighted).

Unseen elements are still being highlighted, however I don't know how to fix this as DOM.isVisible is failing to recognize invisible nodes as invisible.

**Comment by glacambre** [comment]
I forgot to implement incsearch, but you're right, with the current approach it would need to be implemented in the completion mechanism and we'd hit the completion bug again.

**Comment by glacambre** [comment]
I've implemented incsearch but searching for words that are frequently used in a page is slow (for example on the Battle of Mosul wikipedia page it takes almost ten seconds to jump to the word "mosul"). I don't know how to fix this.

Also, as predicted, the same problem as for completions is triggered. I think it'll be fixed when https://github.com/cmcaine/tridactyl/issues/732 is fixed.

**Comment by glacambre** [comment]
`:tutor` needs to be updated with information about find mode before this is merged.

**Comment by bovine3dom** [comment]
I'm happy to merge this without incsearch or regex; we can just leave it unbound by default.

People can bind it to `/` if they want - the bug that was blocking that was fixed ages ago.

Is the main issue remaining moving stuff to content?

**Comment by glacambre** [comment]
I think there are several reasons to wait before merging this. It requires a new permission and we haven't found a way to smoothly upgrade Tridactyl yet. Also, if I remember correctly, incsearch didn't work too well because of https://github.com/cmcaine/tridactyl/issues/732. Moving everything to the content script would indeed probably help with that.

**Comment by bovine3dom** [comment]
@glacambre since #708 is 100% totally fixed now, I'm happy for you to merrily add as many permissions as you want.

Does that unblock this?

File: src/static/css/content.css [hunk 1]
```diff
@@ -50,3 +50,9 @@
         display: none !important;
     }
 }
+
+.TridactylSearchHighlight {
+    opacity: 0.5 !important;
```
Comment (bovine3dom) [inline src/static/css/content.css]: This is a really ugly hack. Nice work ;)

**Comment by bovine3dom** [comment]
Highlighted elements aren't focused (they probably should be?). Apparently using `/` to find a link and then click on it is something I do a lot...

File: src/static/css/content.css (hunk 1 above)
Comment (glacambre) [reply src/static/css/content.css]: Is it? I don't remember what any of this does but it doesn't seem too shocking right now.

Also, don't bother reading any of this code. I tried to rebase it and it was awful, I should probably just rewrite everything and copy/paste where I can.
Sorry it's taking so long.

Comment (glacambre) [reply src/static/css/content.css]: Oh, I think I see what you mean. This is actually the recommended way of doing this on MDN.

Comment (bovine3dom) [reply src/static/css/content.css]: I merged master into it because rebases are for masochists. It was fine. : /

**Comment by bovine3dom** [comment]
The search completions include the place where you've typed the phrase into the command line, which I guess is technically correct, but a little surprising.

**Comment by bovine3dom** [comment]
I'm tempted to unbind `/` and `?`, `n` and `N`, document it all a bit, then merge this as it is.

Do you know of any reason why I shouldn't do that?

**Comment by glacambre** [comment]
Nope, I really don't remember anything about this PR :).

**Pull request merged**